### ISSUE TITLE
Reduce type inflation by decoupling combinator types

### ISSUE THREAD

**Issue by p-e-w** [zesterer/chumsky#35]
Hi there! [I'm writing a computer algebra system](https://github.com/p-e-w/savage) and I thought I'll give Chumsky a try. Your API design is excellent and in general I found Chumsky quite straightforward to use. Thank you for this crate!

That being said, I encountered a major problem: [My simple expression parser](https://github.com/p-e-w/savage/blob/master/savage_core/src/parse.rs) takes almost five minutes to compile and results in a 1.3 GB debug executable. Rust-Analyzer is brought to its knees and barely capable of inspecting the code at all, but reveals that the final `disjunction` parser has a type that is dozens of *pages* when written out!

The main difference between that parser and your "nano_rust" example is that there are a few more levels of precedence, but that's about it. The language being parsed is by no means complicated or syntactically ambiguous. It appears that some mechanism in Chumsky results in exponential growth of parser complexity when nesting parsers hierarchically.

Any idea what's going on here? Is this a bug, or am I just using Chumsky incorrectly?

**Comment by zesterer** [comment]
Unfortunately, Rust deals with nested types extremely poorly. I believe that the work Rust does for a type is at least quadratic over the type's written length, so this can quickly inflate.

Thankfully, there is a solution!

All Chumsky parsers have a [`.boxed()`](https://docs.rs/chumsky/0.5.0/chumsky/trait.Parser.html#method.boxed) method that performs type erasure, cutting down the size of your types significantly (at the cost of a small amount of runtime overhead in the form of a vtable lookup).

I tend to use `.boxed()` at various opportune points to avoid my parser types inflating too much. You can see this happening [here](https://github.com/zesterer/tao/blob/tao2/syntax/src/parse.rs#L280-L560).

If it's any consolation, Rust's iterator API suffers from similar problems when chaining very large numbers of iterators together.

**Comment by p-e-w** [comment]
Thank you for the suggestion, that makes perfect sense. I did notice the `boxed` method as I glanced through the docs, but I guess I skipped reading the full explanation which already covers my problem.

But while it's perhaps unsurprising that the compiler struggles with types that are Megabytes of text when written out, I still don't understand why this would result in a huge artifact being generated. Rust generics are implemented via type erasure at compilation time, right? Seems like the complexity of the type's generic expansion should have no effect on binary size then.
